### PR TITLE
[Merged by Bors] - feat: port Data.Stream.Init

### PR DESCRIPTION
**Pull request by siddhartha-gadgil**
e574b1a4e891376b0ef974b926da39e05da12a06

Due to name collisions with `Stream` in Lean 4, the structure is `Stream'` as earlier defined. Recursive proofs were changed a bit to show termination.

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
Should the file `Data.Vector` be moved to `Init.Data.Vector`. Is it a port of a Lean3 core file?

**Comment by digama0** [comment]
No, unless it is moved in lean 3 as well.

File: Mathlib/Data/Stream/Init.lean [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+universe u v w
+
+namespace Stream'
+
+variable {α : Type u} {β : Type v} {δ : Type w}
+
+instance {α} [Inhabited α] : Inhabited (Stream' α) :=
+  ⟨Stream'.const default⟩
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
```
Comment (semorrison) [inline Mathlib/Data/Stream/Init.lean:38]:
```suggestion
```
and below.

File: Mathlib/Data/Stream/Init.lean [hunk 2]
```diff
@@ -0,0 +1,795 @@
...
+  intro n s₁ s₂
+  match n, s₁, s₂ with
+  |0, s₁, s₂ => rfl
+  |n + 1, s₁, s₂ =>
+    change nth (s₁ ⋈ s₂) (succ (succ (2 * n + 1))) = nth s₂ (succ n)
+    rw [nth_succ, nth_succ, interleave_eq, tail_cons, tail_cons]
+    have : n < succ n := Nat.lt_succ_self n
+    let ih := nth_interleave_right n (tail s₁) (tail s₂)
+    rw [ih]
+    rfl
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:439]:
```suggestion
theorem nth_interleave_right : ∀ (n : Nat) (s₁ s₂ : Stream' α),
    nth (s₁ ⋈ s₂) (2 * n + 1) = nth s₂ n
  | 0, s₁, s₂ => rfl
  | n + 1, s₁, s₂ => by
    change nth (s₁ ⋈ s₂) (succ (succ (2 * n + 1))) = nth s₂ (succ n)
    rw [nth_succ, nth_succ, interleave_eq, tail_cons, tail_cons, 
      nth_interleave_right n (tail s₁) (tail s₂)]
    rfl
```

Comment (kbuzzard) [reply Mathlib/Data/Stream/Init.lean:439]: (this is what the autoporter says, I'm a bit confused about why you changed it)

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Data/Stream/Init.lean [hunk 3]
```diff
@@ -0,0 +1,795 @@
...
+  intro n s₁ s₂
+  match n, s₁, s₂ with
+  |0, s₁, s₂ => rfl
+  |n + 1, s₁, s₂ =>
+    change nth (s₁ ⋈ s₂) (succ (succ (2 * n))) = nth s₁ (succ n)
+    rw [nth_succ, nth_succ, interleave_eq, tail_cons, tail_cons]
+    have : n < succ n := Nat.lt_succ_self n
+    let ih := nth_interleave_left n (tail s₁) (tail s₂)
+    rw [ih]
+    rfl
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:425]: You don't need the `have` and you can just add `nth_interleave_left n (tail s₁) (tail s₂)` to the `rw`

File: Mathlib/Data/Stream/Init.lean [hunk 4]
```diff
@@ -0,0 +1,795 @@
...
+
+theorem head_even (s : Stream' α) : head (even s) = head s :=
+  rfl
+#align stream.head_even Stream'.head_even
+
+theorem tail_even (s : Stream' α) : tail (even s) = even (tail (tail s)) := by
+  rw [even, corec, even, corec]
+  rw [tail_map]
+  apply congrArg <| map (fun s ↦ head s)
+  rw [tail_iterate]
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:462]:
```suggestion
theorem tail_even (s : Stream' α) : tail (even s) = even (tail (tail s)) := by
  unfold even;
  rw [corec_eq]; 
  rfl
```
This is what the autoporter suggests and it works fine. Why did you change it?

File: Mathlib/Data/Stream/Init.lean [hunk 5]
```diff
@@ -0,0 +1,795 @@
...
+theorem nth_even : ∀ (n : Nat) (s : Stream' α), nth (even s) n = nth s (2 * n) := by
+  intro n s
+  match n, s with
+  | 0, s => rfl
+  | succ n, s =>
+    change nth (even s) (succ n) = nth s (succ (succ (2 * n)))
+    rw [nth_succ, nth_succ, tail_even];
+    have : n < succ n := Nat.lt_succ_self n
+    let ih := nth_even n (tail (tail s))
+    rw [ih, nth, nth, tail]
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:504]:
```suggestion
theorem nth_even : ∀ (n : Nat) (s : Stream' α), nth (even s) n = nth s (2 * n) 
  | 0, s => rfl
  | succ n, s => by
    change nth (even s) (succ n) = nth s (succ (succ (2 * n)))
    rw [nth_succ, nth_succ, tail_even, nth_even n]; rfl
```

File: Mathlib/Data/Stream/Init.lean [hunk 6]
```diff
@@ -0,0 +1,795 @@
...
+theorem append_append_stream : ∀ (l₁ l₂ : List α) (s : Stream' α),
+    l₁ ++ l₂ ++ₛ s = l₁ ++ₛ (l₂ ++ₛ s) := by
+  intro l₁ l₂ s
+  match l₁, l₂, s with
+  | [], l₂, s => rfl
+  | List.cons a l₁, l₂, s =>
+    rw [List.cons_append, Stream'.cons_append_stream]
+    let _ := append_append_stream l₁ l₂ s
+    apply congrArg (Stream'.cons a)
+    assumption
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:538]:
```suggestion
theorem append_append_stream : ∀ (l₁ l₂ : List α) (s : Stream' α),
    l₁ ++ l₂ ++ₛ s = l₁ ++ₛ (l₂ ++ₛ s)
  | [], l₂, s => rfl
  | List.cons a l₁, l₂, s => by
    rw [List.cons_append, cons_append_stream, cons_append_stream, append_append_stream l₁]
```

File: Mathlib/Data/Stream/Init.lean [hunk 7]
```diff
@@ -0,0 +1,795 @@
...
+theorem map_append_stream (f : α → β) :
+    ∀ (l : List α) (s : Stream' α), map f (l ++ₛ s) = List.map f l ++ₛ map f s := by
+  intro l s
+  match l, s with
+  | [], s => rfl
+  | List.cons a l, s =>
+    rw [cons_append_stream, List.map_cons, map_cons, cons_append_stream]
+    let _ := map_append_stream f l s
+    apply congrArg (Stream'.cons _)
+    assumption
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:551]:
```suggestion
theorem map_append_stream (f : α → β) :
    ∀ (l : List α) (s : Stream' α), map f (l ++ₛ s) = List.map f l ++ₛ map f s
  | [], s => rfl
  | List.cons a l, s => by
    rw [cons_append_stream, List.map_cons, map_cons, cons_append_stream, map_append_stream f l]
```

File: Mathlib/Data/Stream/Init.lean [hunk 8]
```diff
@@ -0,0 +1,795 @@
...
+#align stream.map_append_stream Stream'.map_append_stream
+
+theorem drop_append_stream : ∀ (l : List α) (s : Stream' α), drop l.length (l ++ₛ s) = s
+  | [], s => by rfl
+  | List.cons a l, s => by
+    rw [List.length_cons, drop_succ, cons_append_stream, tail_cons]
+    let _ := drop_append_stream l s
+    assumption
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:559]:
```suggestion
theorem drop_append_stream : ∀ (l : List α) (s : Stream' α), drop l.length (l ++ₛ s) = s
  | [], s => by rfl
  | List.cons a l, s => by
    rw [List.length_cons, drop_succ, cons_append_stream, tail_cons, drop_append_stream l s]
```

File: Mathlib/Data/Stream/Init.lean [hunk 9]
```diff
@@ -0,0 +1,795 @@
...
+
+theorem nth_take_succ : ∀ (n : Nat) (s : Stream' α),
+    List.nth (take (succ n) s) n = some (nth s n) := by
+  intro n s
+  match n, s with
+  | 0, s => rfl
+  | n + 1, s =>
+    rw [take_succ, add_one, List.nth]
+    let _ := nth_take_succ n (tail s)
+    assumption
```
Comment (kbuzzard) [inline Mathlib/Data/Stream/Init.lean:603]:
```suggestion
theorem nth_take_succ : ∀ (n : Nat) (s : Stream' α),
    List.nth (take (succ n) s) n = some (nth s n) 
  | 0, s => rfl
  | n + 1, s => by
    rw [take_succ, add_one, List.nth, nth_take_succ n]; rfl
```

**Review by kbuzzard — APPROVED** [review]
This is great. A bunch of auto-ported proofs didn't work for you because `rw [def]` seems to behave differently in Lean 4, in the middle of a proof involved in a definition of `def`. I looked at your workarounds and fed some additional information into the rewrites and got them working again; as a result I could golf a bunch of your proofs back to versions which far more closely approximate the autoported versions. Not a big deal but I thought you might be interested! Whether or not you add these golfs, this PR is ready to merge.

File: Mathlib/Data/Stream/Init.lean (hunk 3 above)
Comment (siddhartha-gadgil) [reply Mathlib/Data/Stream/Init.lean:423]: Done. Thanks.

**Comment by jcommelin** [comment]
bors merge

**Pull request closed without merging**